File: android/android/src/main/java/com/supermap/rnsupermap/JSDataDownloadService.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.supermap.data.Dataset;
import com.supermap.data.DatasetVector;
import com.supermap.data.Datasource;
import com.supermap.services.DataDownloadService;

import java.util.Calendar;
import java.util.Map;

public class JSDataDownloadService extends JSServiceBase {
    public static final String REACT_CLASS = "JSDataDownloadService";

    public JSDataDownloadService(ReactApplicationContext context) {
        super(context);
    }

    public static DataDownloadService getObjFromList(String id) {
        return (DataDownloadService)m_ServiceBaseList.get(id);
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    public static String registerId(DataDownloadService obj) {
        for (Map.Entry entry : m_ServiceBaseList.entrySet()) {
            if (obj.equals(entry.getValue())) {
                return (String) entry.getKey();
            }
        }

        Calendar calendar = Calendar.getInstance();
        String id = Long.toString(calendar.getTimeInMillis());
        m_ServiceBaseList.put(id, obj);
        return id;
    }

    @ReactMethod
    public void createObj(String url,Promise promise){
        try{
            DataDownloadService dataDownloadService = new DataDownloadService(url);
            String dataDownloadServiceId = registerId(dataDownloadService);

            WritableMap map = Arguments.createMap();
            map.putString("_dataDownloadServiceId_",dataDownloadServiceId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void download(String dataDownloadServiceId,String fullUrl,int fromIndex,int toIndex,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            dataDownloadService.download(fullUrl,fromIndex,toIndex);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void downloadByName(String dataDownloadServiceId,String serviceName,String datasourceName,
            String datasetName,int fromIndex,int toIndex,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            dataDownloadService.download(serviceName,datasourceName,datasetName,fromIndex,toIndex);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void downloadAll(String dataDownloadServiceId,String fullUrl,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            dataDownloadService.downloadAll(fullUrl);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void downloadAllByName(String dataDownloadServiceId,String serviceName,String datasourceName,
                               String datasetName,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            dataDownloadService.downloadAll(serviceName,datasourceName,datasetName);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void downloadDataset(String dataDownloadServiceId,String urlDatset,String datasourceId,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            Datasource datasource = JSDatasource.m_DatasourceList.get(datasourceId);
            dataDownloadService.downloadDataset(urlDatset,datasource);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void updateDataset(String dataDownloadServiceId,String urlDatset,String datasetId,Promise promise){
        try{
            DataDownloadService dataDownloadService = getObjFromList(dataDownloadServiceId);
            Dataset dataset = JSDataset.getObjById(datasetId);
            dataDownloadService.updateDataset(urlDatset,(DatasetVector)dataset);

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }
}

File: ios/RNObjCode/JSDataDownloadService.m
//
//  JSDataDownloadService.m
//  HelloWorldDemo
//
//  Created by 王子豪 on 2016/11/21.
//  Copyright © 2016年 Facebook. All rights reserved.
//

#import "JSDataDownloadService.h"
#import "SuperMap/DataDownloadService.h"
#import "JSObjManager.h"
@implementation JSDataDownloadService
RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(createObj,url:(NSString*)url resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  DataDownloadService* DDS = [[DataDownloadService alloc]init];
  if(DDS){
    NSInteger key = (NSInteger)DDS;
    [JSObjManager addObj:DDS];
    resolve(@{@"_dataDownloadServiceId_":@(key).stringValue});
  }else{
    reject(@"dataDownloadService",@"create dataDownloadService failed!!!",nil);
  }
}

RCT_REMAP_METHOD(download,DDSId:(NSString*)DDSId fullURL:(NSString*)fullURL fromIndex:(NSInteger)fromIndex toIndex:(NSInteger)toIndex resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    DataDownloadService* DDS = [JSObjManager getObjWithKey:DDSId];
    FeatureSet* set = [DDS downLoadUrl:fullURL fromIndex:(int)fromIndex toIndex:(int)toIndex];
  if(set){
    resolve(@"done");
  }else{
    reject(@"dataDownloadService",@"download failed!!!",nil);
  }
}

RCT_REMAP_METHOD(downloadByName,DDSId:(NSString*)DDSId serviceName:(NSString*)serviceName datasourceName:(NSString*)datasourceName datasetName:(NSString*)datasetName fromIndex:(int)fromIndex toIndex:(int)toIndex resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    DataDownloadService* DDS = [JSObjManager getObjWithKey:DDSId];
//    [DDS download:<#(NSString *)#> serviceName:<#(NSString *)#> datasourceName:<#(NSString *)#> datasetName:<#(NSString *)#> fromIndex:<#(int)#> toIndex:<#(int)#>]
}

RCT_REMAP_METHOD(downloadAll,DDSId:(NSString*)DDSId URL:(NSString*)URL resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    DataDownloadService* DDS = [JSObjManager getObjWithKey:DDSId];
    FeatureSet* set = [DDS downLoadUrlAll:URL];
    if(set){
      resolve(@"done");
    }else{
      reject(@"dataDownloadService",@"download failed!!!",nil);
    }
}

RCT_REMAP_METHOD(downloadAllByName,DDSId:(NSString*)DDSId serviceName:(NSString*)serviceName datasourceName:(NSString*)datasourceName datasetName:(NSString*)datasetName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
  ////url
}

RCT_REMAP_METHOD(downloadDataset,DDSId:(NSString*)DDSId urlDatset:(NSString*)urlDatset datasourceId:(NSString*)datasourceId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    DataDownloadService* DDS = [JSObjManager getObjWithKey:DDSId];
    Datasource* datasource = [JSObjManager getObjWithKey:datasourceId];
    [DDS downloadDatasetFrom:urlDatset toDatasource:datasource];
}

RCT_REMAP_METHOD(updateDataset,DDSId:(NSString*)DDSId urlDatset:(NSString*)urlDatset datasetId:(NSString*)datasetId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    DataDownloadService* DDS = [JSObjManager getObjWithKey:DDSId];
    DatasetVector* dataset = [JSObjManager getObjWithKey:datasetId];
    [DDS updateDatasetFrom:urlDatset toDatasetVector:dataset];
}
@end
